File: backend/src/app/modules/Jobs/job.model.ts
import { Schema, model } from "mongoose";
import { TJob } from "./job.interface";

const jobSchema = new Schema<TJob>(
  {
    title: {
      type: String,
      required: [true, "Job title is required"],
      trim: true,
    },
    company: {
      type: String,
      required: [true, "Company name is required"],
      trim: true,
    },
    location: {
      type: String,
      required: [true, "Job location is required"],
      trim: true,
    },
    requiredSkills: {
      type: [String],
      required: [true, "Required skills are mandatory"],
      validate: {
        validator: (arr: string[]) => arr.length > 0,
        message: "At least one skill is required",
      },
    },
    experienceLevel: {
      type: String,
      required: [true, "Experience level is required"],
      enum: ["Fresher", "Junior", "Mid", "Senior"],
    },
    jobType: {
      type: String,
      required: [true, "Job type is required"],
      enum: ["Internship", "Part-time", "Full-time", "Freelance"],
    },
    description: {
      type: String,
      trim: true,
    },
    applyLink: {
      type: String,
      trim: true,
    },
    careerTrack: {
      type: String,
      enum: [
        "Web Development",
        "Data",
        "Design",
        "Marketing",
        "Cybersecurity",
        "AI/ML",
        "Mobile App",
        "Other",
      ],
    },
    postedAt: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true, 
  }
);

export const Job = model<TJob>("Job", jobSchema);

File: lib/models/job_model.dart
// models/job_model.dart
class JobModel {
  final String id;
  final String title;
  final String company;
  final String location;
  final List<String> requiredSkills;
  final String experienceLevel;
  final String jobType;
  final String description;
  final String applyLink;
  final String careerTrack;
  final DateTime postedAt;
  int matchPercentage = 0;

  JobModel({
    required this.id,
    required this.title,
    required this.company,
    required this.location,
    required this.requiredSkills,
    required this.experienceLevel,
    required this.jobType,
    required this.description,
    required this.applyLink,
    required this.careerTrack,
    required this.postedAt,
  });

  factory JobModel.fromJson(Map<String, dynamic> json) {
    return JobModel(
      id: json['_id'] ?? '',
      title: json['title'] ?? 'Untitled Job',
      company: json['company'] ?? 'Unknown Company',
      location: json['location'] ?? 'Location not specified',
      requiredSkills: List<String>.from(json['requiredSkills'] ?? []),
      experienceLevel: json['experienceLevel'] ?? 'Not specified',
      jobType: json['jobType'] ?? 'Full-time',
      description: json['description'] ?? '',
      applyLink: json['applyLink'] ?? '',
      careerTrack: json['careerTrack'] ?? 'Other',
      postedAt: DateTime.tryParse(json['postedAt'] ?? '') ?? DateTime.now(),
    );
  }
}
